Add `Controller` interface.

haydenheroux opened the pull request.

haydenheroux committed:
Add  interface.

lddtwo456 commented:
I love the structure of this, it's gonna take some work to allow it to replace our current controllers but the way this is structured is really cool. Also as long as we don't use any hardware from other vendors and stick to just TalonFX controlled motors we shouldn't run into any complications with vendor specific control requests or StatusSignal equivalents that'll make this kind of structure very messy and wordy, so this seems very doable. Some stuff we're gonna need to consider to fit this in our codebase are probably:

- More useful types for **ValuesType**, one could be called something like **MotorValues**, so we can get the same amount of information from the motor that we do with our current system (position, velocity, acceleration, voltage, stator current, supply current)
- I'd prefer to keep subsystem specific logic inside the code for a subsystem, like conversion from angle to distance, checking if you're at the target setpoint, and calculating trajectories, I think it lessens confusion for other teams helping debug our code, as well as team members learning how our code works. I tried with the current rewrite to keep essentially just PID and feedforward gains to the controller, everything else in the subsystem
- Passing a **VelocityController** to the **ProfiledPositionController** means less code repetition, but it also might get a bit awkward when you're defining a **ProfiledPositionController** to have to also make a **VelocityController** beforehand
- We need support for multiple different motor outputs, like single and leader follower type stuff, maybe with a **MotorOutput** class defined with CAN ids that will be passed into the **VelocityController** (and like I said probably **ProfiledPositionController**) constructor rather than a **TalonFX** object, and it could just take control requests as input and send those control requests to however many motors are using the same output

Oh and I also forgot to add that i think your usage of a velocity controller inside a position controller also doesnt allow for any position based PID at all for a profiled mechanism? Which like yeah ideally you shouldnt need any, but there could be cases where a bit of PID can help, or situations where you want both the ability to have it follow a motion profile and also use pure PID to snap to say a moving setpoint, and i think having a position controller that takes a position and velocity setpoint and directly sets voltages of some **MotorOutput** class could mean a little less code redundancy, and make things a bit less complicated at the cost of being a little less cool.

haydenheroux committed:
Introduce OutputValuesType.

The second parameter represents sensor values associated with the output. In a future commit, there will be a generic Output interface that will return OutputValues.

haydenheroux commented:
Added the first useful type for **ValuesType** (now **OutputValuesType**, since there will be an **Output** type in the future), **TalonFXOutputValues** for the TalonFX motor controller used by Krakens.

haydenheroux committed:
Add Output interface.

haydenheroux commented:
Added **Output** interface. Next I am working on a complete example with the **Elevator**.

haydenheroux commented:
Here's a block diagram that represents how the classes relate:
<img width="2487" height="1186" alt="image" src="https://github.com/user-attachments/assets/229d3cd5-0070-4769-968b-381fd0c137ec" />

lddtwo456 commented:
Seems pretty good to me, though since we're using just **TalonFX** controlled motors couldn't we assume that all types of motor outputs will be controllable with **ControlRequest**s and will have **TalonFXOutputValues** as their **ValuesType**? Then we could assume all implementations of **Output<ControlInputType, OutputValuesType>** will be **Output<ControlRequest, TalonFXOutputValues>**, and get rid of the type parameters for some nice simplicity, while also allowing any kind of **ControlRequest** to be used without needing to make a bunch of implementations like **VoltageOutputTalonFX**, **VoltageOutputTalonFX2**, **PositionVoltageOutputTalonFX**, **PositionVoltageOutputTalonFX2**, etc.

I also think that naming-wise having two completely different types defined with Output in the name is a little weird, maybe we could change **Output** to **MotorOutput** and **MotorOutputValues** to just **MotorValues** or something like that.

Also I still think it'd be better to keep all feedforward and simulation inside subsystem code, and have more general controller implementations with both open and closed loop control that can essentially be disabled just by having 0 feedback or 0 feedforward gains, at least to start. Like a position controller implementation with pid and ff that takes **PositionVelocity** as a setpoint or something. I think having that kind of subsystems specific trajectory generation and mechanism simulation inside subsystem code would make it a bit easier to follow the motor control code and make things feel a bit less magical and confusing for people new to the codebase, and also mean less very similar implementations like **OpenLoopPositionController**, **ClosedLoopPositionController**, **PositionController** that type of stuff.

haydenheroux commented:
I think that having the **ControlInputType** of every **Output** being **ControlRequest** would cause issues for simulation because we'd have to unwrap the **ControlRequest** first. I think that is the opposite of how we should do it; the **Controller** should be generating physically meaningful values (like a voltage or current), and it should be the responsibility of the **Output**s downstream to convert that to an action (simulation update or setting the motor's **ControlRequest**.

I completely agree about **Output**, the naming is pretty tricky and I'm not really happy with the name. If we used control theory terminology, **Output** would be **Plant** or **System** and **OutputValues** would be **ProcessVariables**. **MotorOutputValues** to **MotorValues** is a good shout though, I just kept **Output** in the name because it "implemented" the type.

I don't think keeping feedforward in the subsystem is a good idea because we might want to use built-in motor control, then we'd have dead code in the subsystem (or, we would have to toggle with an if statement).

I somewhat agree on keeping trajectory / motion profile generation in the subsystem, but for the simple cases (trapezoidal and exponential profiles), the TalonFX can do that natively. That's why I put **MotionProfile** behind the **Controller** interface; because TalonFXs can be used as a **Controller** or **Output** directly, and we want to keep the same abstraction.

haydenheroux committed:
Rename MotorOutputValues to MotorValues.

haydenheroux committed:
Reorganize packages.

haydenheroux committed:
Semi-working example.

haydenheroux committed:
Working example of profiled open-loop. Previous commit contains closed-loop.

lddtwo456 commented:
I think reducing abstraction as much as we can other than abstraction that allows us to do things like not care if we're controlling a real or a fake motor or 2 motors at the same time is generally a good idea for our robot code both to reduce confusion of other teams looking at our code and make it easier for new team members to learn how our code works. Honestly, even removing the distinction between a position controller and motor controller might be fine for now, in favor of just having a simple MotorOutput interface that is given control requests and gives back sensor values (that's maybe an unnecessary amount of unabstraction I don't really know right now). 

Generally though I think less abstraction means some code repetition but also means that we can optimize the control of each individual subsystem where it's necessary, instead of spending time making a cool expandable library that allows you to make implementations for perfectly controlled elevators and pivots and rollers only to make like one or two cool implementations that we actually use and leave the rest of the cool structure dormant just taking up unnecessary space. Leaving subsystem specific control code to subsystem code means less unnecessary complexity, avoids the problem of creating YAGSL for motor control and abstracting away all the things that are important for future programmers to learn, and makes our code easier to follow and more easily improvable. We can bring back some abstraction as we go, and also the cool YAGSL but for motor control library could be a cool side project that would be good for learning and maybe being a utility we do use in comp code in the future idk

haydenheroux committed:
Add two-motor output.

haydenheroux committed:
Refactor ClosedLoopPositionController to include feedforward voltage and update Elevator to use new controller.

haydenheroux committed:
Clean up.
